Section Feature Extraction (#161)
johko/computer-vision-course: chapters/en/Unit 1 - Fundamentals/feature-extraction/feature_description.mdx

﻿# Feature Description

## How Can We Represent Features In Data Structures?

Representing features in data is crucial for organizing and manipulating data effectively. Features, or attributes or variables, can be diverse, ranging from numerical values and categories to more complex structures like images or text. Some ways to represent features for computer vision tasks are:

- **Numerical features**

  - Arrays/Lists: Simplest form to store numerical values. Each element in the array corresponds to a feature.
  - Tensors: Multidimensional arrays often used in the machine learning frameworks to handle large sets of numerical data efficiently.

- **Categorical features**

  - Dictionaries/Lists: Assigning categories to numerical labels or directly storing categorical values.

>   - One hot encoding:\*\* Transforming categorical variables into binary vectors where each bit represents a category.

Reviewer: ```suggestion
  - One hot encoding: Transforming categorical variables into binary vectors where each bit represents a category.
```

2. **Distinctiveness:** A good descriptor captures unique information about the object. It should be able to discriminate between different objects or parts of an image and be distinct enough to differentiate them from similar elements.
3. **Dimensionality:** Good descriptors often have a manageable size, conveying enough information without being excessively large. Balancing dimensionality is crucial for efficiency in processing and storage.
4. **Locality:** Descriptors often identify local features within an image. Local descriptors focus on specific regions or keypoints and describe the characteristics of these areas, enabling matching and recognition of similar regions across different images.
5. **Repeatability:** Descriptors should be consistent and reproducible across multiple instances of the same object or scene even in the presence of noise or minor variations.

> 6. **Compatibility with Matching Algorithms:** Descriptors are often used in conjunction with matching algorithms to find correspondences between different images. A good descriptor should be suitable for the matching algorithm being used, whether it's based on distance metrics, machine learning models, or other techniques.

Reviewer: ```suggestion
- **Compatibility with Matching Algorithms:** Descriptors are often used in conjunction with matching algorithms to find correspondences between different images. A good descriptor should be suitable for the matching algorithm being used, whether it's based on distance metrics, machine learning models, or other techniques.
```

7. **Computational Efficiency:** Efficiency is really a crucial part of a descriptor, especially in real-time applications. Descriptors should be computationally feasible for quick processing, particularly in scenarios where speed is crucial, such as robotics or autonomous vehicles.
8. **Adaptability:** Descriptors that can adapt or learn from the data they are processing can be highly effective, especially in situations where the characteristics of the objects or scenes may change over time. This can increase the useability of the descriptor.

> 9. **Noise Robustness:** Descriptors should be able to handle noise in the image data without significantly compromising their ability to accurately represent the underlying features.

Reviewer: ```suggestion
- **Noise Robustness:** Descriptors should be able to handle noise in the image data without significantly compromising their ability to represent the underlying features accurately.
```

## Some of the Techniques Used in Feature Descriptors

### SIFT

![Basic Working of SIFT](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/feature-extraction-feature-matching/Original-SIFT-algorithm-flow.png)

It stands for scale invariant feature transform. It is a widely used algorithm in computer vision and image processing for detecting and describing local features in images.

The working of SIFT is given below:

> 1. **Scale Space Extrema detection:** It starts by detecting potential interest points in an image across multiple scales. It looks for locations in the image where the difference of Gaussian function reaches a maximum or minimum over space and scale. These keypoint locations are considered stable under various scale changes.

Reviewer: ```suggestion
- **Scale Space Extrema detection:** It starts by detecting potential interest points in an image across multiple scales. It looks for locations in the image where the difference of Gaussian function reaches a maximum or minimum over space and scale. These keypoint locations are considered stable under various scale changes.
```

> 2. **Keypoint Localization:** Once potential keypoints are identified, SIFT refines their positions to sub-pixel accuracy and discards low-contrast keypoints and keypoints on edges to ensure accurate localization.

Reviewer: ```suggestion
- **Keypoint Localization:** Once potential keypoints are identified, SIFT refines their positions to sub-pixel accuracy and discards low-contrast keypoints and keypoints on edges to ensure accurate localization.
```

> 3. **Orientation Assignment:** SIFT computes a dominant orientation for each keypoint based on local image gradient directions. This step makes the descriptor invariant to image rotation.

Reviewer: ```suggestion
- **Orientation Assignment:** SIFT computes a dominant orientation for each keypoint based on local image gradient directions. This step makes the descriptor invariant to image rotation.
```

> 4. **Descriptor Generation:** A descriptor is computed for each keypoint region, capturing information about the local image gradients in the vicinity of the keypoint. This descriptor is a compact representation that encapsulates the key characteristics of the image patch surrounding the keypoint.

Reviewer: ```suggestion
- **Descriptor Generation:** A descriptor is computed for each keypoint region, capturing information about the local image gradients near the keypoint. This descriptor is a compact representation that encapsulates the key characteristics of the image patch surrounding the keypoint.
```

> 5. **Descriptor Matching:** Finally, these descriptors are used for matching keypoints between different images. The descriptors from one image are compared to those in another image to find correspondences.

Reviewer: ```suggestion
- **Descriptor Matching:** Finally, these descriptors are used for matching keypoints between different images. The descriptors from one image are compared to those in another image to find correspondences.
```

SIFT's robustness to various image transformations and its ability to find distinctive features in an image makes it valuable in applications like object recognition, image stitching, and 3D reconstruction.

You can learn more about SIFT using the following references:

> 1. [Introduction to SIFT (Scale-Invariant Feature Transform)](https://docs.opencv.org/4.x/da/df5/tutorial_py_sift_intro.html)

Reviewer: ```suggestion
- [Introduction to SIFT (Scale-Invariant Feature Transform)](https://docs.opencv.org/4.x/da/df5/tutorial_py_sift_intro.html)
```

> 2. https://www.educative.io/answers/what-is-sift

Reviewer: ```suggestion
[What is SIFT](https://www.educative.io/answers/what-is-sift)
```

> 3. https://www.cse.iitb.ac.in/~ajitvr/CS763/SIFT.pdf

Reviewer: ```suggestion
- [SIFT](https://www.cse.iitb.ac.in/~ajitvr/CS763/SIFT.pdf)
```

### SURF

![Basic Working of SURF](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/feature-extraction-feature-matching/Flow-Chart-for-SURF-Feature-Detection.png)

It stands for Speeded Up Robust features. It is another popular algorithm in computer vision and image processing. It is particularly known for its speed and robustness in detecting and describing local image features.

The basic workflow of SURF is given below:

> 1. **Integral images:** SURF utilizes integral images, precomputed representations of the original image. They allow fast calculations of rectangular area sums within an image, enabling a quicker feature computation.

Reviewer: ```suggestion
- **Integral images:** SURF utilizes integral images, precomputed representations of the original image. They allow fast calculations of rectangular area sums within an image, enabling a quicker feature computation.
```

> 2. **Blob detection:** Similar to other feature detection algorithms, SURF starts by identifying potential interest points or keypoints in the image. It uses a Hessian matrix to detect blobs or regions that exhibit significant variations in intensity in multiple directions and scales. These regions are potential keypoints.

Reviewer: ```suggestion
- **Blob detection:** Similar to other feature detection algorithms, SURF starts by identifying potential interest points or keypoints in the image. It uses a Hessian matrix to detect blobs or regions that exhibit significant variations in intensity in multiple directions and scales. These regions are potential keypoints.
```

> 3. **Scale Selection**: It determines the scale of the keypoints by identifying regions with significant changes in scale-space. It analyzes the determinant of the Hessian matrix across different scales to find robust keypoints at multiple scales.

Reviewer: ```suggestion
- **Scale Selection**: It determines the scale of the keypoints by identifying regions with significant changes in scale-space. It analyzes the determinant of the Hessian matrix across different scales to find robust keypoints at multiple scales.
```

> 4. **Orientation Assignment:** For each detected keypoint, SURF assigns a dominant orientation. This is done by calculating Haar wavelet responses in different directions around the keypoint’s neighborhood. Unlike SIFT, SURF uses a set of rectangular filters (Haar wavelets) applied to subregions of the keypoint's neighborhood. The responses of these filters are used to create a feature vector representing the keypoint.

Reviewer: ```suggestion
- **Orientation Assignment:** For each detected keypoint, SURF assigns a dominant orientation. This is done by calculating Haar wavelet responses in different directions around the keypoint’s neighborhood. Unlike SIFT, SURF uses a set of rectangular filters (Haar wavelets) applied to subregions of the keypoint's neighborhood. The responses of these filters are used to create a feature vector representing the keypoint.
```

> 5. **Descriptor Matching:** The generated descriptors are then used for matching keypoints between different images. Matching involves comparing the feature vectors of keypoints in one image to those in another image to find correspondences.

Reviewer: ```suggestion
- **Descriptor Matching:** The generated descriptors are then used to match key points between different images. Matching involves comparing the feature vectors of keypoints in one image to those in another image to find correspondences.
```

The key strengths of SURF lie in its computational efficiency, which is achieved through the use of integral images and Haar wavelet approximations while maintaining robustness to scale, rotation, and illumination changes. This makes SURF suitable for real-time applications where speed plays a crucial part in object detection, tracking, and image stitching.

You can learn more about SURF using the following references:

- [Medium Blogpost - Introduction to SURF ](https://medium.com/@deepanshut041/introduction-to-surf-speeded-up-robust-features-c7396d6e7c4e)

>   2.[OpenCV Tutorial - Introduction to SURF (Speeded-Up Robust Features)] (https://docs.opencv.org/3.4/df/dd2/tutorial_py_surf_intro.html)

Reviewer: ```suggestion
 - [OpenCV Tutorial - Introduction to SURF (Speeded-Up Robust Features)] (https://docs.opencv.org/3.4/df/dd2/tutorial_py_surf_intro.html)
```

> 3. [Journal Paper - Feature Extraction Using SURF Algorithm for Object Recognition](https://www.ijtra.com/view/feature-extraction-using-surf-algorithm-for-object-recognition.pdf)

Reviewer: ```suggestion
- [Journal Paper - Feature Extraction Using SURF Algorithm for Object Recognition](https://www.ijtra.com/view/feature-extraction-using-surf-algorithm-for-object-recognition.pdf)
```
